Add best practices doc (#556)
grafana/plugin-tools: docusaurus/docs/create-a-plugin/develop-a-plugin/best-practices.md

---
id: best-practices
title: Best practices
description: An index of best practices for plugins.
keywords:
  - grafana
  - plugins
  - plugin
  - best practices
sidebar_position: 5
---

> This document summarizes best practices for Grafana plugins. This live document is updated as new best practices are identified.

Reviewer: ```suggestion
Learn how to build high quality plugins for Grafana through this regularly updated collection of best practices.

Something missing from the list? [Let us know](https://github.com/grafana/plugin-tools/issues/new).
```

## General

> - Choose the right plugin type for your use case: panel, data-source or app. See [Plugin types](/introduction/plugin-types-usage) for more information.

Reviewer: Maybe too basic for a "best practice"? Or is this about reminding users to understand the fundamentals - if so we could reword.

Author: yeah, I wanted to start from the basics, agree that it's not that useful though, should I remove it?

Reviewer: We could have something like "Never forget the fundamentals, be sure to read up on [introductory](/introduction) concepts for plugin development. 

(Haven't checked link syntax)

Author: that's a bit too vague, I can just remove it

> - Verify that your data-source or app plugin can be provisioned. See [Provisioning](https://grafana.com/docs/grafana/latest/administration/provisioning/#data-sources) for more information.

Reviewer: Is there something particular that a plugin dev needs to do to ensure provisioning works as expected?

Author: Difficult to say, we have found bugs when testing plugins when a provisioned data-source does not render properly properly in the config editor for example or when upgrading a plugin, when a provisioned data source for one version does not work in the next. Not sure if we can be more specific here. I just wanted to point out that provisioning is a thing and plugin devs should be aware of that.

Reviewer: Oh interesting. By all means keep it - would be great if we could narrow down the cause/resolution though

> - Include default dashboards in your data-source or app plugin. See [Bundling of dashboards](https://grafana.com/docs/grafana/latest/administration/provisioning/#data-sources) for more information.

Reviewer: Maybe we want to link to https://grafana.com/developers/plugin-tools/introduction/plugin-types-usage#bundling-of-dashboards?

Author: wrong link, good catch :+1:

- Ensure that the minimum version of Grafana is correct. This should point to the earliest version of Grafana that your plugin fully supports. This is specified as `grafanaDependency` in your plugin.json.

## Panel plugins

- Do not store or use credentials. Panel plugins do not have a way to securely store credentials. If your plugin needs to use credentials, consider using a data-source plugin instead and a panel plugin to display the information returned by the data-source.

> - See how you can add more advanced options to your panel plugin [here](/create-a-plugin/extend-a-plugin/custom-panel-option-editors).

Reviewer: This feels like it depends on the panel rather than being a best practice which could be generally applied by developers?

Author: I thought this to be a bit of feature discovery as well. Maybe someone is doing something cumbersome with the pre-defined panel options because they don't know you can do custom ones. WDYT?

Reviewer: That's fair. How about something like "If the default panel options are not a good fit for what you're trying to offer users, consider creating custom options"

## Data-source plugins

- If your plugin needs to store credentials, use `secureJsonData` instead of `jsonData`. The former is encrypted at rest while the latter is not. See [Secure JSON data](/create-a-plugin/extend-a-plugin/add-authentication-for-data-source-plugins#store-configuration-in-securejsondata) for more information.
- Implement a query builder. This is highly useful for users who are not familiar with the query language of the data-source. See, for example, the [Query Builder for Microsoft SQL Server](https://grafana.com/docs/grafana/latest/datasources/mssql/query-editor/#builder-mode) which helps write SQL queries for that service.

> - Add a health check for your plugin. This is used to verify that the data-source is working properly. More information in [Health checks](/introduction/backend-plugins#health-checks). Depending on the type of plugin (frontend or backend) the implementation methods differ. See this [example](https://github.com/grafana/grafana-plugin-examples/blob/5441fe2f818e28cdeb06eb7066ff198dd34bb0ab/examples/datasource-http/src/DataSource.ts#L81-L115) for a health check in the frontend or this [example](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b23645251997088ac7a1707a72d3fd9248/examples/datasource-basic/pkg/plugin/datasource.go#L77) in the backend.

Reviewer: ```suggestion
- Add a health check for your plugin. This is used to verify that the data-source is working properly. How this is implemented depends on whether the plugin has a backend. See our examples for health checks in the [frontend](https://github.com/grafana/grafana-plugin-examples/blob/5441fe2f818e28cdeb06eb7066ff198dd34bb0ab/examples/datasource-http/src/DataSource.ts#L81-L115) and [backend](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b23645251997088ac7a1707a72d3fd9248/examples/datasource-basic/pkg/plugin/datasource.go#L77). Read more about [Health checks](/introduction/backend-plugins#health-checks).
```

- Add [dashboard variables](https://grafana.com/docs/grafana/latest/dashboards/variables/) support. Dashboard (or template) variables allow users to create dynamic dashboards. There are two aspects of adding support for variables in your plugin. The first is allowing querying your data-source and return values that will be used as variables. The second is replacing existing variables in other queries. See how you can do it [here](/create-a-plugin/extend-a-plugin/add-support-for-variables#add-support-for-query-variables-to-your-data-source). Pay special attention when selecting "All values" since it may require specific logic to join variable values.
- Add annotations support. Annotations allow users to add contextual information to their dashboards and it's possible to use queries to define them. See how you can do it [here](/create-a-plugin/extend-a-plugin/enable-for-annotations).

> - When building frontend components, make sure to use [Grafana components](https://developers.grafana.com/ui/latest/index.html?path=/docs/docs-overview-intro--page) as the base and follow the [Saga Design System](https://grafana.com/blog/2023/11/07/saga-design-system-shaping-the-future-of-user-experiences-at-grafana-labs/).

Reviewer: Only curious - why link to the blog rather than direct to the design system?

Author: It was the first entry in Google and explains a little bit more the why :smile: but it's true that the overview page will be kept more up to date.

- Add query editor help. Query editors can be complex and it's useful to provide help to users. See how you can do it [here](/create-a-plugin/extend-a-plugin/add-query-editor-help).
- Skip hidden or empty queries. This avoids executing unnecessary or wrong queries. See this [example](https://github.com/grafana/grafana/blob/fd5f66083c91b9759ae7772f99b80c9342b93290/public/app/plugins/datasource/loki/datasource.ts#L1085).
- Specify a default query. This can be useful for users to discover how queries are written for the plugin. See an [example](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b/examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/src/datasource.ts#L39-L41).

> - Avoid using `console.log` in your plugin.

Reviewer: why?

Author: it's a best practice, in general: https://eslint.org/docs/latest/rules/no-console. It can clutter the browser console and hide real errors.

### Frontend (only) plugins

> - Frontend-only data sources typically use the [Grafana proxy](/create-a-plugin/extend-a-plugin/add-authentication-for-data-source-plugins#add-a-proxy-route-to-your-plugin) to access an external service. While this is a simple way of adding support for queries in your plugin (and doesn't require Golang knowledge), it's generally recommended to write a backend part for your plugin. The use cases can be found [here](/introduction/backend-plugins#use-cases-for-implementing-a-backend-plugin).

Reviewer: ```suggestion
- Frontend-only data sources typically use the [Grafana proxy](/create-a-plugin/extend-a-plugin/add-authentication-for-data-source-plugins#add-a-proxy-route-to-your-plugin) to access an external service. While this is a simple way of adding support for queries in your plugin (and doesn't require Golang knowledge), it's generally recommended to write a backend part for your plugin. The additional use cases this unlocks can be found [here](/introduction/backend-plugins#use-cases-for-implementing-a-backend-plugin).
```

### Backend plugins

- Add support for alerts. Backend plugins have inherent support for alerts but it needs to be enabled. Simply add `"enabled": true` to your plugin.json file.
- Use the `CallResourceHandler` interface to serve custom HTTP requests. More information in [Resource handlers](/introduction/backend-plugins#resource-handlers). This is useful, for example, when providing query builders. See an example [here](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b23645251997088ac7a1707a72d3fd9248/examples/app-with-backend/pkg/plugin/app.go#L35).

> - Add metrics to your data-source. More information [here](/introduction/backend-plugins#collect-metrics).

Reviewer: we'll need to remember to update these when Marcus' PR merges

Author: which PR?

Reviewer: https://github.com/grafana/plugin-tools/pull/547

Author: got it, the one in the next line is the one that will break.

Author: seems like that PR is still undergoing, we can update this doc afterwards.

- Add tracing to your data-source. In addition to metrics, tracing allows you to deep dive into your plugin performance. More information [here](/create-a-plugin/extend-a-plugin/add-distributed-tracing-for-backend-plugins). And an example [here](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b/examples/datasource-http-backend/pkg/plugin/datasource.go#L141-L156).
- Keep cached connections. This is an important optimization. More information [here](/introduction/backend-plugins#caching-and-connection-pooling) and an example [here](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b23645251997088ac7a1707a72d3fd9248/examples/datasource-http-backend/pkg/plugin/datasource.go#L40-L66).

> - Add macro support. Macros are similar to variables but they are tipically evaluated in the backend and it allows to return values based on environment data like the current time selection. These are usually defined using the syntax `$__macroName` (e.g. `$__timeFilter`). See this [example](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b23645251997088ac7a1707a72d3fd9248/examples/datasource-basic/pkg/query/macro.go) to discover how you can implement it.

Reviewer: ```suggestion
- Add macro support. Macros are similar to variables, but they are typically evaluated in the backend and can return values based on environment data like the current time selection. These are usually defined using the syntax `$__macroName` (e.g. `$__timeFilter`). See this [example](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b23645251997088ac7a1707a72d3fd9248/examples/datasource-basic/pkg/query/macro.go) to discover how you can implement support.
```

- Do not use the local file system. Different plugins share the same environment and for security reasons, plugins should not rely on local files.
- Similar to the above, avoid the usage of environment variables. Those are also a security risk and should be avoided. For data-source-specific configuration, use the `jsonData` or `secureJsonData` fields in the plugin.json file. In case some generic configuration is needed for the plugin (shared among data-sources), use the [`plugin` configuration](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#pluginplugin_id).
- Plugins should not execute arbitrary code in the backend. Again, this is a security risk and should be avoided. If your plugin needs to execute code, provide a list of allowed commands and validate the input before executing it.
- Logged arguments should not expose sensitive information, such as secrets.
- In general, any error happening should be logged with level `error`.

> - Using the `info` level is normally not recommended: use the `debug` level instead.

Reviewer: Need to check Marcus' PR on this too just to make sure we're aligned on our advice.

Author: same, which PR? seems to be a a few

Reviewer: Same one https://github.com/grafana/plugin-tools/pull/547

Author: there is a ["Best practices"](https://github.com/grafana/plugin-tools/pull/547/files#diff-0625ec7b29cd40a4e5940aa569122d446643d559b6090e5810c57d378d1136b8R169) there, we can just link to that :+1:

> ## App plugins

Reviewer: It would be good to reference UI extensions here, and maybe even a slight advertisement for apps in general in that they allow for more tailored monitoring experience and for user actions.

Author: can you draft a suggestion for this? so I make sure what you are thinking

Reviewer: Consider whether contributing a UI extension would help a user appropriately discover your app and continue a given workflow. Additionally, if your app provides context which could be used in other apps - creating an extension point would allow these apps to do so, with no further changes required in your app.

Author: Added the "Apps" item to the "General Section". Do we have a link we can add there so people know more about Apps?

## Panel plugins

- Do not store or use credentials. Panel plugins do not have a way to securely store credentials. If your plugin needs to use credentials, consider using a data-source plugin instead and a panel plugin to display the information returned by the data-source.
- See how you can add more advanced options to your panel plugin [here](/create-a-plugin/extend-a-plugin/custom-panel-option-editors).

## Data-source plugins

- If your plugin needs to store credentials, use `secureJsonData` instead of `jsonData`. The former is encrypted at rest while the latter is not. See [Secure JSON data](/create-a-plugin/extend-a-plugin/add-authentication-for-data-source-plugins#store-configuration-in-securejsondata) for more information.
- Implement a query builder. This is highly useful for users who are not familiar with the query language of the data-source. See, for example, the [Query Builder for Microsoft SQL Server](https://grafana.com/docs/grafana/latest/datasources/mssql/query-editor/#builder-mode) which helps write SQL queries for that service.

> - Add a health check for your plugin. Depending on the type of plugin (frontend or backend) the implementation methods differ, see the following sections for more information.

Reviewer 2: Should there be a link here to the specific section perhaps?

Author: I thought that but it's not that helpful either since it was difficult to find. I ended up moving everything to this line so there are no references to other sections.

---
id: best-practices
title: Best practices
description: An index of best practices for plugins.
keywords:
  - grafana
  - plugins
  - plugin
  - best practices
sidebar_position: 0.5
---

# Best practices for plugin development

> We have compiled a list of good practices for building high-quality plugins for Grafana. We recommend referring to them when building your own plugin for best results.

Reviewer 3: ```suggestion
We have compiled a list of practices for building high-quality plugins for Grafana that we have found beneficial. Refer to them when building your own plugin for best results.
```

- **Skip hidden or empty queries.** This avoids executing unnecessary or wrong queries. Data sources implementing [`DataSourceWithBackend`](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b23645251997088ac7a1707a72d3fd9248/examples/datasource-basic/src/datasource.ts#L5) only need to implement the method `filterQuery`. Refer to this [example](https://github.com/grafana/grafana/blob/fd5f66083c91b9759ae7772f99b80c9342b93290/public/app/plugins/datasource/loki/datasource.ts#L1085).
- **Specify a default query.** Default queries can help users to discover how queries are written for the plugin. Refer to this [example](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b/examples/datasource-streaming-backend-websocket/streaming-backend-websocket-plugin/src/datasource.ts#L39-L41).
- **Avoid using `console.log` in your plugin.** Console messages are usually for debugging purposes and therefore not suitable to ship to the client.

### Frontend (only) plugins

> - **Frontend-only data sources typically use the [Grafana proxy](/create-a-plugin/extend-a-plugin/add-authentication-for-data-source-plugins#add-a-proxy-route-to-your-plugin) to access an external service.** While this is a simple way of adding support for queries in your plugin (and doesn't require Golang knowledge), there are use cases for which writing a backend plugin is necessary. Refer to [Backend plugins](/introduction/backend-plugins#use-cases-for-implementing-a-backend-plugin) for more information about those.

Reviewer 3: ```suggestion
- **Data sources running only on the frontend typically use the [Grafana proxy](/create-a-plugin/extend-a-plugin/add-authentication-for-data-source-plugins#add-a-proxy-route-to-your-plugin) to access an external service.** This s a simple way of adding support for queries in your plugin, and it doesn't require Golang knowledge. However, there are use cases for which writing a backend plugin is necessary. Refer to [Backend plugins](/introduction/backend-plugins#use-cases-for-implementing-a-backend-plugin) for more information about those.
```

### Backend plugins

- **Add support for alerting.** Backend plugins have inherent support for [Grafana Alerting](https://grafana.com/docs/grafana/latest/alerting/) but this support needs to be enabled. Simply add `"enabled": true` to your `plugin.json` file.
- **Use the `CallResourceHandler` interface to serve custom HTTP requests.** For more information, refer to [Resource handlers](/introduction/backend-plugins#resource-handlers). This is useful, for example, when providing query builders, as shown in this [example](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b23645251997088ac7a1707a72d3fd9248/examples/app-with-backend/pkg/plugin/app.go#L35).

> - **Add metrics to your data source.** More information in [Backend plugins](/introduction/backend-plugins#collect-metrics).

Reviewer 3: ```suggestion
- **Add metrics to your data source.** Find more information in [Backend plugins](/introduction/backend-plugins#collect-metrics).
```

- **Add macro support.** Macros are similar to variables, but they are typically evaluated in the backend and can return values based on environment data like the current time selection. It can be useful, for example, to evaluate alerts during a dynamic period. These are usually defined using the syntax `$__macroName` (for example, `$__timeFilter`). Refer to this [example](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b23645251997088ac7a1707a72d3fd9248/examples/datasource-basic/pkg/query/macro.go) to discover how you can implement support. Some pre-defined macros are available in the [plugin SDK `macros` package](https://github.com/grafana/grafana-plugin-sdk-go/tree/main/experimental/macros).
- **For SQL data sources, refer to the [`sqlutil` package in the SDK](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go/data/sqlutil).** It includes multiple helpers to work with SQL data sources for things like data frame conversion, default macros, and so on. Also, consider using the [`sqlds` package](https://pkg.go.dev/github.com/grafana/sqlds) which highly simplifies the implementation of SQL data sources.

> - **Don't use the local file system.** Different plugins share the same environment. For security reasons, plugins should not rely on local files.

Reviewer 3: ```suggestion
- **Don't use the local file system.** Different plugins share the same environment. For security reasons, plugins shouldn't rely on local files.
```

> - **Don't use environment variables.** Environment variables are also a security risk and should be avoided. For configuration to a particular data source, use the `jsonData` or `secureJsonData` fields in the `plugin.json` file. In case some generic configuration is needed for the plugin (shared among data sources), use the [`plugin` configuration](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#pluginplugin_id).

Reviewer 3: ```suggestion
- **Don't use environment variables.** Environment variables are also a security risk and should be avoided. For configuration to a particular data source, use the `jsonData` or `secureJsonData` fields in the `plugin.json` file. If configuration is needed for the plugin which is shared among data sources, then use the [`plugin` configuration](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#pluginplugin_id).
```

- **Plugins should not execute arbitrary code in the backend.** Again, this is a security risk and should be avoided. If your plugin needs to execute code, provide a list of allowed commands and validate the input before executing it.
- **Don't expose sensitive information.** For security reasons, avoid exposing sensitive information such as secrets.
- **In general, any error happening should be logged with level `error`.**
- **Don't use the `info` level: use the `debug` level instead.**

## App plugins

- **Specify a root page for your app.** If your app defines multiple pages, make sure to select a default one that will be used as a landing page for your plugin.
- **To generate dynamic apps, consider using [Grafana Scenes](https://grafana.com/developers/scenes/).**

> - **Consider whether contributing a [UI extension](https://grafana.com/developers/plugin-tools/ui-extensions/) would help a user appropriately discover your app and continue a given workflow.** Additionally, if your app provides context that be used in other apps, then create an extension point to allow these apps to do so, with no further changes required in your app.

Reviewer 3: ```suggestion
- **Consider  contributing a [UI extension](https://grafana.com/developers/plugin-tools/ui-extensions/)**. UUI extensions can help a user to discover your app in context and continue a given workflow. Additionally, if your app provides context that be used in other apps, then create an extension point to allow these apps to do so, with no further changes required in your app.
```

---
id: best-practices
title: Best practices
description: An index of best practices for plugins.
keywords:
  - grafana
  - plugins
  - plugin
  - best practices
sidebar_position: 0.5
---

# Best practices for plugin development

We have compiled a list of good practices for building high-quality plugins for Grafana. We recommend referring to them when building your own plugin for best results.

Is something missing from this list? [Let us know](https://github.com/grafana/plugin-tools/issues/new).

## General

- **Verify that your data source or app plugin can be provisioned.** Refer to [Provisioning](https://grafana.com/docs/grafana/latest/administration/provisioning/#data-sources) for more information.
- **Include default dashboards in your data source or app plugin.** Refer to [Bundling of dashboards](https://grafana.com/developers/plugin-tools/introduction/plugin-types-usage#bundling-of-dashboards) for more information.
- **Ensure that the minimum version of Grafana is correct.** Make sure that the `grafanaDependency` in your `plugin.json` points to the earliest version of Grafana that your plugin fully supports.

> - **Consider building an [app plugin](/introduction/plugin-types-usage#app-plugins).** App plugins offer more flexibility and additional capabilities than panel or data source plugins, allowing you to build more tailored monitoring experiences for a given service. Apps provide a better user experience than simply including default dashboards with a data source.

Reviewer: >When reading this I get the impression that I should always build an app rather than panel or datasource. Or is that we (you) want?

That's valid feedback, it's perhaps too strongly worded. The intent was just to highlight that apps do offer a lot more in terms of capabilities and to encourage developers, particularly say ISVs, that if they take advantage of those capabilities then they can make it significantly easier for users to interact with their service.

🤔 of a rewrite

> - **Consider building an [app plugin](/introduction/plugin-types-usage#app-plugins).** App plugins offer more flexibility and additional capabilities than panel or data source plugins, allowing you to build more tailored monitoring experiences for a given service. Apps provide a better user experience than simply including default dashboards with a data source.

Reviewer: wdyt about:

```
**Consider building an [app plugin](/introduction/plugin-types-usage#app-plugins).**

If your users would benefit from more of an opinionated experience then an app plugin may be an appropriate choice, offering both increased flexibility and additional capabilities like custom pages and UI extensions.
```

> - **Consider building an [app plugin](/introduction/plugin-types-usage#app-plugins).** App plugins offer more flexibility and additional capabilities than panel or data source plugins, allowing you to build more tailored monitoring experiences for a given service. Apps provide a better user experience than simply including default dashboards with a data source.

Reviewer: If this doesn't solve the concern, remove this bullet to unblock the PR

> - **Consider building an [app plugin](/introduction/plugin-types-usage#app-plugins).** App plugins offer more flexibility and additional capabilities than panel or data source plugins, allowing you to build more tailored monitoring experiences for a given service. Apps provide a better user experience than simply including default dashboards with a data source.

Author: "opinionated experience" seems a bit vage for me. Since it's not a very specific best-practice lets just remove it. This info should be already in the intro page for plugin types.

> - **Consider building an [app plugin](/introduction/plugin-types-usage#app-plugins).** App plugins offer more flexibility and additional capabilities than panel or data source plugins, allowing you to build more tailored monitoring experiences for a given service. Apps provide a better user experience than simply including default dashboards with a data source.

Reviewer: Go for it

- **Add annotations support.** Annotations allow users to add contextual information to their dashboards and it's possible to use queries to define them. For more information, refer to [Enable annotations](/create-a-plugin/extend-a-plugin/enable-for-annotations).
- **Practice good front-end design.** When building frontend components, make sure to use [Grafana components](https://developers.grafana.com/ui/latest/index.html?path=/docs/docs-overview-intro--page) as the base and follow the [Saga Design System](https://grafana.com/developers/saga/About/overview).
- **Add query editor help.** Query editors can be complex and it's useful to provide help to users. For more information, refer to [Add query editor help](/create-a-plugin/extend-a-plugin/add-query-editor-help).

> - **Skip hidden or empty queries.** This avoids executing unnecessary or wrong queries. Refer to this [example](https://github.com/grafana/grafana/blob/fd5f66083c91b9759ae7772f99b80c9342b93290/public/app/plugins/datasource/loki/datasource.ts#L1085).

Reviewer 4: Okay sorry you're right. But meant if you implement filterquery for backend datasource queries are automatically filtered. If frontend only you have to do the filtering all by yourself in the query function.

- **Add macro support.** Macros are similar to variables, but they are typically evaluated in the backend and can return values based on environment data like the current time selection. These are usually defined using the syntax `$__macroName` (for example, `$__timeFilter`). Refer to this [example](https://github.com/grafana/grafana-plugin-examples/blob/0532f8b23645251997088ac7a1707a72d3fd9248/examples/datasource-basic/pkg/query/macro.go) to discover how you can implement support.
- **For SQL data sources, refer to the [`sqlutil` package in the SDK](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go/data/sqlutil).** It includes multiple helpers to work with SQL data sources for things like data frame conversion, default macros, and so on.
- **Don't use the local file system.** Different plugins share the same environment. For security reasons, plugins should not rely on local files.

> - **Don't use environment variables.** Environment variables are also a security risk and should be avoided. For configuration to a particular data source, use the `jsonData` or `secureJsonData` fields in the `plugin.json` file. In case some generic configuration is needed for the plugin (shared among data sources), use the [`plugin` configuration](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#pluginplugin_id).

Author: mmh, that doc doesn't say much? Just "Use the Grafana configuration file to configure your app [plugin_id](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#pluginplugin_id).". I think https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#pluginplugin_id has a little bit more of information and examples?

> - **Don't use environment variables.** Environment variables are also a security risk and should be avoided. For configuration to a particular data source, use the `jsonData` or `secureJsonData` fields in the `plugin.json` file. In case some generic configuration is needed for the plugin (shared among data sources), use the [`plugin` configuration](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#pluginplugin_id).

Reviewer 4: I thought it would be nice since that in turn link to the grafana config docs.

> - **Don't use environment variables.** Environment variables are also a security risk and should be avoided. For configuration to a particular data source, use the `jsonData` or `secureJsonData` fields in the `plugin.json` file. In case some generic configuration is needed for the plugin (shared among data sources), use the [`plugin` configuration](https://grafana.com/docs/grafana/latest/setup-grafana/configure-grafana/#pluginplugin_id).

Author: It doesn't show any extra information so we can avoid the extra click :shrug:

- **Plugins should not execute arbitrary code in the backend.** Again, this is a security risk and should be avoided. If your plugin needs to execute code, provide a list of allowed commands and validate the input before executing it.
- **Don't expose sensitive information.** For security reasons, avoid exposing sensitive information such as secrets.
- **In general, any error happening should be logged with level `error`.**
- **Don't use the `info` level: use the `debug` level instead.**

## App plugins

> - **Ensure that your app can be enabled and disabled.**

Author: Well, this is something provided by Grafana so really not something the user really needs to verify. Removing it.

> - **Specify a default page for your app.**

Author: Added some more specific comment to specify that this only applies if the app has multiple pages.
